### PR TITLE
add capsa IAlbumAccessorSession and more capsa calls

### PR DESCRIPTION
**Pull request by HookedBehemoth**

### CODE REVIEW

**Comment by yellows8** [comment]
Changes pending then?

**Comment by HookedBehemoth** [comment]
Not to the functionality

File: nx/source/services/capsa.c [hunk 1]
```diff
@@ -191,3 +197,101 @@ Result capsaResetAlbumMountStatus(CapsAlbumStorage storage) {
...
+    return serviceDispatchInOut(&g_capsaAccessor, 2007, in, *actual_size,
+        .buffer_attrs = { SfBufferAttr_HipcMapAlias | SfBufferAttr_Out },
+        .buffers = { { buffer, size } },
+    );
+}
+
+Result capsaOpenAlbumMovieStream(u64 *stream, const CapsAlbumFileId *file_id) {
+    Result rc=0;
+
+    if (!serviceIsActive(&g_capsaAccessor)) rc =_capsaOpenAccessorSessionForApplication(&g_capsaAccessor);
```
Comment (yellows8) [inline nx/source/services/capsa.c]: `= _capsaOpenAccessorSessionForApplication(`

File: nx/source/services/capsa.c [hunk 2]
```diff
@@ -191,3 +197,101 @@ Result capsaResetAlbumMountStatus(CapsAlbumStorage storage) {
 Result capsaRefreshAlbumCache(CapsAlbumStorage storage) {
     return _capsaCmdInU8NoOut(&g_capsaSrv, storage, 8011);
 }
+
+static Result _capsaOpenAccessorSessionForApplication(Service* srv_out) {
```
Comment (yellows8) [inline nx/source/services/capsa.c]: Sysver check missing.

File: nx/source/services/capsa.c [hunk 3]
```diff
@@ -238,58 +238,82 @@ static Result _capsaReadImageDataFromAlbumMovieReadStream(u64 stream, s64 offset
 }
 
 Result capsaOpenAlbumMovieStream(u64 *stream, const CapsAlbumFileId *file_id) {
+    if (hosversionBefore(4,0,0))
+        return MAKERESULT(Module_Libnx, LibnxError_IncompatSysVer);
+
     Result rc=0;
 
-    if (!serviceIsActive(&g_capsaAccessor)) rc =_capsaOpenAccessorSessionForApplication(&g_capsaAccessor);
+    if (!serviceIsActive(&g_capsaAccessor)) rc =_capsaOpenAccessorSession(&g_capsaAccessor);
```
Comment (yellows8) [inline nx/source/services/capsa.c]: Again, add whitespace after `=`.

File: nx/source/services/capsa.c [hunk 4]
```diff
@@ -238,58 +238,82 @@ static Result _capsaReadImageDataFromAlbumMovieReadStream(u64 stream, s64 offset
...
 
     if (R_SUCCEEDED(rc)) rc = _capsaOpenAlbumMovieReadStream(stream, file_id);
 
     return rc;
 }
 
 Result capsaCloseAlbumMovieStream(u64 stream) {
+    if (hosversionBefore(4,0,0))
```
Comment (yellows8) [inline nx/source/services/capsa.c]: Mention sysver in the .h for these then?

Comment (HookedBehemoth) [reply nx/source/services/capsa.c]: Isn't the one above capsaOpenAlbumMovieStream enough until real descriptions follow?

File: nx/source/services/capsa.c (hunk 3 above)
Comment (HookedBehemoth) [reply nx/source/services/capsa.c]: Misunderstood your issue.

File: nx/source/services/capsa.c (hunk 4 above)
Comment (yellows8) [reply nx/source/services/capsa.c]: sure

**Comment by HookedBehemoth** [comment]
Added the missing documentation for all of caps:a.

File: nx/include/switch/services/capsa.h [hunk 5]
```diff
@@ -18,31 +18,268 @@ void capsaExit(void);
...
+ * @param[in] storage \ref CapsAlbumStorage
+ * @param[out] count Amount of files.
+ */
 Result capsaGetAlbumFileCount(CapsAlbumStorage storage, u64* count);
+
+/**
+ * @brief Gets a listing of \ref CapsAlbumEntry, where the AlbumFile's storage matches the input one.
+ * @param[in] storage \ref CapsAlbumStorage
+ * @param[out] count Total output entries.
+ */
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: Missing params in docs.

File: nx/include/switch/services/capsa.h [hunk 6]
```diff
@@ -18,31 +18,268 @@ void capsaExit(void);
...
+ * @brief Gets the mount status of the specified \ref CapsAlbumStorage.
+ * @param[in] storage \ref CapsAlbumStorage
+ * @param[out] is_mounted
+ */
 Result capsaIsAlbumMounted(CapsAlbumStorage storage, bool* is_mounted);
+
+/**
+ * @brief Returns the AlbumUsage for a specified \ref CapsAlbumStorage.
+ * @param[in] storage \ref CapsAlbumStorage
+ * @param[out] out \ref CapsAlbumUsage
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: CapsAlbumUsage2

File: nx/include/switch/services/capsa.h [hunk 7]
```diff
@@ -18,31 +18,268 @@ void capsaExit(void);
...
+ * @param[in] storage \ref CapsAlbumStorage
+ * @param[out] count Total output entries.
+ */
 Result capsaGetAlbumFileList(CapsAlbumStorage storage, u64* count, CapsAlbumEntry* buffer, u64 size);
+
+/**
+ * @brief Loads a file into the specified buffer.
+ * @param[in] file_id \ref CapsAlbumFileId
+ * @param[out] out_size Size of the AlbumFile.
+ * @param[out] workbuf File output buffer.
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: Is this buffer cleared or not?(update comments depending on this) Same for the equivalent buffers elsewhere.

File: nx/include/switch/services/capsa.h [hunk 8]
```diff
@@ -18,31 +18,268 @@ void capsaExit(void);
...
+ * @brief Copies an AlbumFile to the specified \ref CapsAlbumStorage.
+ * @param[in] file_id \ref CapsAlbumFileId
+ * @param[in] dst_storage \ref CapsAlbumStorage
+ */
 Result capsaStorageCopyAlbumFile(const CapsAlbumFileId *file_id, CapsAlbumStorage dst_storage);
+
+/**
+ * @brief Gets the mount status of the specified \ref CapsAlbumStorage.
+ * @param[in] storage \ref CapsAlbumStorage
+ * @param[out] is_mounted
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: Missing comment.

File: nx/include/switch/services/capsa.h [hunk 9]
```diff
@@ -18,31 +18,268 @@ void capsaExit(void);
...
+ * @brief Returns the result for a AlbumStorage mount.
+ * @param[in] storage \ref CapsAlbumStorage
+ */
 Result capsaGetAlbumMountResult(CapsAlbumStorage storage);
+
+/**
+ * @brief Returns the AlbumUsage for a specified \ref CapsAlbumStorage.
+ * @note Only available on [4.0.0+].
+ * @param[in] storage \ref CapsAlbumStorage
+ * @param[out] out \ref CapsAlbumUsage
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: CapsAlbumUsage16

File: nx/include/switch/services/capsa.h [hunk 10]
```diff
@@ -18,31 +18,268 @@ void capsaExit(void);
...
+/**
+ * @brief Load the ScreenShotThumbnailImage for the specified AlbumFile.
+ * @note Only available on [4.0.0+].
+ * @param[in] file_id \ref CapsAlbumFileId
+ * @param[in] opts \ref CapsScreenShotDecodeOption
+ * @param[out] image RGBA8 image output buffer.
+ * @param[in] image_size Image buffer size, should be at least large enough for RGBA8 320x180.
+ * @param[out] workbuf Work buffer, cleared to 0 by the cmd before it returns.
+ * @param[in] workbuf_size Work buffer size, must be at least the size of the JPEG within the AlbumFile.
+ * @param[out] out Data buffer.
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: Clarify what this is?

File: nx/include/switch/services/capsa.h [hunk 11]
```diff
@@ -18,31 +18,268 @@ void capsaExit(void);
...
+ * @brief Refreshs Album Cache for the specified AlbumStorage.
+ * @param[in] storage \ref CapsAlbumStorage
+ */
 Result capsaRefreshAlbumCache(CapsAlbumStorage storage);
+
+/**
+ * @brief Opens an AlbumMovieStream.
+ * @note This opens IAlbumAccessorSession if not previously opened, it's closed during \ref capsaExit.
+ * @note Up to 4 streams can be open at the same time. Multiple streams can be open at the same time for the same \ref CapsAlbumFileId.
+ * @note Only available on [4.0.0+]
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: Missing period (same below).

File: nx/source/services/capsa.c [hunk 12]
```diff
@@ -28,24 +28,24 @@ Service* capsaGetServiceSession_Accessor(void) {
...
     u8 inval = storage;
     return serviceDispatchInOut(&g_capsaSrv, 0, inval, *count);
 }
 
-Result capsaGetAlbumFileList(CapsAlbumStorage storage, u64* count, CapsAlbumEntry* buffer, u64 buffer_size) {
+Result capsaGetAlbumFileList(CapsAlbumStorage storage, u64 *count, CapsAlbumEntry *entries, u64 size) {
     u8 inval = storage;
     return serviceDispatchInOut(&g_capsaSrv, 1, inval, *count,
         .buffer_attrs = { SfBufferAttr_HipcMapAlias | SfBufferAttr_Out },
         .buffers = { { buffer, buffer_size } },
```
Comment (yellows8) [inline nx/source/services/capsa.c]: This wasn't updated to match the updated params.

File: nx/include/switch/services/capsa.h [hunk 13]
```diff
@@ -42,7 +44,7 @@ Result capsaGetAlbumFileList(CapsAlbumStorage storage, u64* count, CapsAlbumEntr
  * @param[out] workbuf File output buffer.
  * @param[in] workbuf_size Size of the workbuf.
  */
-Result capsaLoadAlbumFile(const CapsAlbumFileId *file_id, u64 *out_size, void* workbuf, u64 workbuf_size);
+Result capsaLoadAlbumFile(const CapsAlbumFileId *file_id, u64 *out_size, void *workbuf, u64 workbuf_size);
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: Keep `void* `.

File: nx/include/switch/services/capsa.h (hunk 10 above)
Comment (HookedBehemoth) [reply nx/include/switch/services/capsa.h]: I don't know ^^

File: nx/include/switch/services/capsa.h (hunk 7 above)
Comment (HookedBehemoth) [reply nx/include/switch/services/capsa.h]: Bad phrasing I'd say as this isn't a work buffer.

File: nx/include/switch/services/capsa.h (hunk 10 above)
Comment (HookedBehemoth) [reply nx/include/switch/services/capsa.h]: I'll check if I can find something when I get home.

Comment (yellows8) [reply nx/include/switch/services/capsa.h]: Probably the same UserData used by capsu, verify that?

Comment (HookedBehemoth) [reply nx/include/switch/services/capsa.h]: Checked and it's width, height, CapsScreenShotAttribute and then it all zeroes.
If it is CapsLoadAlbumScreenShotImageOutputForApplication it's userdata is empty.

Comment (yellows8) [reply nx/include/switch/services/capsa.h]: It's a fixed-size buffer, so replace out/out_size with {ptr for relevant struct}.

Comment (HookedBehemoth) [reply nx/include/switch/services/capsa.h]: Shall I remove "ForApplication" in this and "CapsScreenShotAttributeForApplication" as it isn't exclusively for Applications?

Comment (yellows8) [reply nx/include/switch/services/capsa.h]: Add 0x450-byte struct `CapsLoadAlbumScreenShotImageOutput` and use that.

**Comment by HookedBehemoth** [comment]
Ninty added 1002 at 3.0.0 and changed it at 4.0.0.
From 3.0.0-3.0.2 it didn't do anything but return 0x4ce.
Should it still be allowed to be called before 4.0.0?

**Comment by yellows8** [comment]
Yeah check for sysver [4.0.0+] for that.

File: nx/include/switch/services/caps.h [hunk 14]
```diff
@@ -175,6 +175,14 @@ typedef struct {
     u8 reserved[0xac];                                    ///< Unused.
 } CapsLoadAlbumScreenShotImageOutputForApplication;
 
+/// LoadAlbumScreenShotImageOutput
+typedef struct {
+    s64 width;                                            ///< Width. Official sw copies this to a s32 output field.
+    s64 height;                                           ///< Height. Official sw copies this to a s32 output field.
+    CapsScreenShotAttributeForApplication attr;           ///< \ref CapsScreenShotAttributeForApplication
+    u8 unk_x50[0x400];
```
Comment (yellows8) [inline nx/include/switch/services/caps.h]: `///< Unused.`

File: nx/include/switch/services/capsa.h [hunk 15]
```diff
@@ -18,31 +18,313 @@ void capsaExit(void);
...
+/**
+ * @brief Load the ScreenShotImage for the specified AlbumFile.
+ * @note Only available on [3.0.0+].
+ * @param[out] width Output image width. Optional, can be NULL.
+ * @param[out] height Output image height. Optional, can be NULL.
+ * @param[out] attr \ref CapsScreenShotAttributeForApplication
+ * @param[in] file_id \ref CapsAlbumFileId
+ * @param[in] opts \ref CapsScreenShotDecodeOption
+ * @param[out] image RGBA8 image output buffer.
+ * @param[in] image_size Image buffer size, should be at least large enough for RGBA8 320x180.
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: This is a ScreenShotImage, so shouldn't this be 1280x720?(same for capsaLoadAlbumScreenShotImageEx1)

**Comment by HookedBehemoth** [comment]
I think I'm done now.
Missing are:
- SaveEditedScreenShot (removed with 3.0.0 so I have no means to actually test it)
- GetAlbumEntryFromApplicationAlbumEntry (can't figure out why it's not working)
- GetAlbumEntryFromApplicationAlbumEntryAruid (didn't even try, see above)
- SetInternalErrorConversionEnabled (seams to only be useful for dev (?))
- LoadMakerNoteInfoForDebug (yeah...)

File: nx/include/switch/services/capsa.h [hunk 16]
```diff
@@ -181,6 +182,52 @@ Result capsaGetAlbumMountResult(CapsAlbumStorage storage);
  */
 Result capsaGetAlbumUsage16(CapsAlbumStorage storage, CapsAlbumUsage16 *out);
 
+/**
+ * @brief Returns the start and end of the Applet Id range.
+ * @note Only available on [6.0.0+].
+ * @param[out] success Returns bool over wether the call was handled or not.
```
Comment (yellows8) [inline nx/include/switch/services/capsa.h]: wether->whether

File: nx/source/services/capsa.c [hunk 17]
```diff
@@ -184,6 +184,68 @@ Result capsaGetAlbumUsage16(CapsAlbumStorage storage, CapsAlbumUsage16 *out) {
...
+    u64 app_ids[2];
+    struct {
+        bool success;
+        u8 pad[0x3];
+    } out;
+    Result rc = serviceDispatchOut(&g_capsaSrv, 18, out,
+        .buffer_attrs = { SfBufferAttr_HipcMapAlias | SfBufferAttr_Out | SfBufferAttr_HipcMapTransferAllowsNonSecure, },
+        .buffers = { { app_ids, sizeof(app_ids) }, },
+    );
+    *min = app_ids[0];
```
Comment (yellows8) [inline nx/source/services/capsa.c]: if (R_SUCCEEDED(rc)) { if (outvar) ...

**Comment by HookedBehemoth** [comment]
*out values

File: nx/include/switch/services/caps.h [hunk 18]
```diff
@@ -175,6 +181,26 @@ typedef struct {
     u8 reserved[0xac];                                    ///< Unused.
 } CapsLoadAlbumScreenShotImageOutputForApplication;
 
+/// LoadAlbumScreenShotImageOutput
+typedef struct {
+    s64 width;                                            ///< Width. Official sw copies this to a s32 output field.
+    s64 height;                                           ///< Height. Official sw copies this to a s32 output field.
+    CapsScreenShotAttributeForApplication attr;           ///< \ref CapsScreenShotAttributeForApplication
```
Comment (yellows8) [inline nx/include/switch/services/caps.h]: Use CapsScreenShotAttribute instead, likewise elsewhere for capsa cmds.

**Pull request merged**